### PR TITLE
Vault CA bugfixes

### PR DESCRIPTION
**Pull request by kisunji**
### Description

Fixing edge case bugs uncovered by https://github.com/hashicorp/consul/issues/19051

### Testing & Reproduction steps

Added tests for goroutine leak detection.

The retry functionality is a revert of #16143 and was tested manually.

### CODE REVIEW

File: agent/connect/ca/provider_vault_test.go [hunk 1]
```diff
@@ -237,8 +238,68 @@ func TestVaultCAProvider_Configure(t *testing.T) {
...
+		}
+		cfg := vaultProviderConfig(t, testVault.Addr, token, config)
+
+		err := provider.Configure(cfg)
+		require.Error(t, err)
+
+		retry.RunWith(retry.TwoSeconds(), t, func(r *retry.R) {
+			profile := pprof.Lookup("goroutine")
+			sb := strings.Builder{}
+			require.NoError(r, profile.WriteTo(&sb, 2))
```
Comment (kisunji) [inline agent/connect/ca/provider_vault_test.go:274]: pprof is a runtime profile which can dump goroutines.
```
profile.WriteTo(&sb, 2)
```
The 2 here means "format it like golang stacktraces".

We are looking for this trace in particular:
```
        goroutine 138 [select]:
        github.com/hashicorp/consul/agent/connect/ca.(*VaultProvider).renewToken(0xc00077bc20, {0x28f2bd8, 0xc0008ce6e0}, 0xc0008bae00)
        	/Users/chriskim/code/consul/agent/connect/ca/provider_vault.go:240 +0x1c5
        created by github.com/hashicorp/consul/agent/connect/ca.(*VaultProvider).Configure in goroutine 135
        	/Users/chriskim/code/consul/agent/connect/ca/provider_vault.go:184 +0x67c
```

File: agent/connect/ca/provider_vault.go [hunk 2]
```diff
@@ -223,6 +230,12 @@ func (v *VaultProvider) renewToken(ctx context.Context, watcher *vaultapi.Lifeti
 	go watcher.Start()
 	defer watcher.Stop()
 
+	retrier := retry.Waiter{
+		MinFailures: 1,
+		MinWait:     1 * time.Second,
+		Jitter:      retry.NewJitter(20),
+	}
+
```
Comment (cthain) [inline agent/connect/ca/provider_vault.go:238]: minor: Were these values arrived at empirically? It may be worth adding a short comment indicating why these values are set the way they are. My initial thought was _"do these need to be configurable?"_. The answer is probably _"No."_, but it would be nice to be certain about that :)

File: agent/connect/ca/provider_vault_test.go (hunk 1 above)
Comment (cthain) [reply agent/connect/ca/provider_vault_test.go:274]: Great job thinking through and implementing a testing solution to this! Not a straightforward, run-of-the-mill unit testing problem :+1:

**Review by cthain — APPROVED** [review]
One non-blocking comment. Nice work on this @kisunji.

**Review by NickCellino — APPROVED** [review]
Take my review with a grain of salt since I'm not familiar with this codebase but looks good to me!

And +1 to @cthain's comments. The test to detect the goroutine leak is super cool. I learned some new stuff!
